### PR TITLE
Add README for the condition variable

### PR DESCRIPTION
**Pull request by OscarShiang**
This commit includes the README file for condition variable and the modification of docs/demos.md.

### CODE REVIEW

File: demos/conditionvariables/README.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+This function initializes the condition variable pointed by `cv`.
+
+Before using functions like `wait()`, `signal()` and `broadcast()`, we must set up the variables with this function.
+
+## `wait`
+
+When we use this function, the thread will be blocked by the condition variable pointed by `cv`.
+
+After calling the function, the thread's `ThreadState` will be changed to `wait`. Since the scheduler only chooses the thread whose state is either `init`, `running` or `suspended`, the blocking thread will keep waiting until other threads signal through the same condition variable.
```
Comment (DavidSpickett) [inline demos/conditionvariables/README.md]: The state is `waiting`.

I would reword this.

"After calling the function, the thread's state will be changed to `waiting`. This means it will remain idle until signalled through the same condition variable."

That way we don't need to list all the states.

Side note: choosing a `running` thread *should* never happen but it's one of those things that probably could happen in some weird circumstance.

File: demos/conditionvariables/README.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+## `signal`
+
+This function unblocks a thread that is blocked on the specified condition variable `cv`.
+
+`signal()` unlocks the thread by setting the state to `suspended`. So the thread can be chosen by the scheduler.
```
Comment (DavidSpickett) [inline demos/conditionvariables/README.md]: You can remove the full stop here, make it one sentence.

File: demos/conditionvariables/README.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+If there exists a thread that can be waked up, the function returns `true`. When there has no thread can be signalled, it returns `false` as a consequence.
```
Comment (DavidSpickett) [inline demos/conditionvariables/README.md]: The meaning here is fine but some phrasing changes would help.

"If there was a waiting thread that was signalled, the function returns `true`. When there are no threads that can be signalled, it returns `false`."

Side note: it would be "woken up" but I admit it also took me a while to work that one out :)

File: demos/conditionvariables/README.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+## `wait`
+
+When we use this function, the thread will be blocked by the condition variable pointed by `cv`.
+
+After calling the function, the thread's `ThreadState` will be changed to `wait`. Since the scheduler only chooses the thread whose state is either `init`, `running` or `suspended`, the blocking thread will keep waiting until other threads signal through the same condition variable.
+
+## `signal`
+
+This function unblocks a thread that is blocked on the specified condition variable `cv`.
```
Comment (DavidSpickett) [inline demos/conditionvariables/README.md]: Clarify this some with:
"This function unblocks the first thread that is blocked on..."

This emphasises the difference between signal and broadcast which is later.

File: demos/conditionvariables/README.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+`signal()` unlocks the thread by setting the state to `suspended`. So the thread can be chosen by the scheduler.
+
+If there exists a thread that can be waked up, the function returns `true`. When there has no thread can be signalled, it returns `false` as a consequence.
+
+## `broadcast`
+
+This function unblocks all threads blocked on the specified condition variable `cv`.
+
+`broadcast()` has a similar functionality to `signal()`. However, when we use `broadcast()`, we will wake up all the threads blocked by the condition variable while `signal()` only wakes up one of them.
```
Comment (DavidSpickett) [inline demos/conditionvariables/README.md]: A little trim to the second sentence:
"`broadcast()` will wake up all the threads blocked by the condition variable while `signal()` only wakes up one of them."

File: demos/conditionvariables/README.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+## Walkthrough
+
+From starter, the `setup()` function creates 5 threads that are set to run `waiter()` and initializes a condition variable called `cond_var`. After the creation, the `setup()` thread jumps to `signaller()` function.
```
Comment (DavidSpickett) [inline demos/conditionvariables/README.md]: You could just start with:
"The `setup()` function..."

It's neater.

File: demos/conditionvariables/README.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+If there exists a thread that can be waked up, the function returns `true`. When there has no thread can be signalled, it returns `false` as a consequence.
+
+## `broadcast`
+
+This function unblocks all threads blocked on the specified condition variable `cv`.
+
+`broadcast()` has a similar functionality to `signal()`. However, when we use `broadcast()`, we will wake up all the threads blocked by the condition variable while `signal()` only wakes up one of them.
+
+## Walkthrough
+
```
Comment (DavidSpickett) [inline demos/conditionvariables/README.md:41]: I see you're ignoring the misc. tests setup does at the very start. Which I think is a good way to go. 👍 

It's obvious they are kinda standalone things and they don't get in the way of the actual demo.

File: demos/conditionvariables/README.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+From starter, the `setup()` function creates 5 threads that are set to run `waiter()` and initializes a condition variable called `cond_var`. After the creation, the `setup()` thread jumps to `signaller()` function.
+
+In `signaller()`, it first yields to check if the condition variable successfully blocks all other threads. Then the `signaller` thread uses `signal()` to wake 2 threads. The scheduler will choose the thread waked up by `cond_var` after the `signaller` thread yields.
```
Comment (DavidSpickett) [inline demos/conditionvariables/README.md]: "threads woken up by"

threads plural

Worth mentioning when the other threads do their wait. Something like:
"Each `waiter` thread will do some work and then wait on `cond_var` so we'll end up back in `signaller`. Then the `signaller` thread..."

I would be more clear what happens here. It's not that we signal 2 threads then yield and they run one after another, its signaller, thread, signaller, thread. (maybe that's what you're saying but it's a bit unclear).

Maybe:
"Then the `signaller` thread uses `signal()` to first wake and then yield to, 2 threads."

That makes the order clearer.

File: demos/conditionvariables/README.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+Once thread 1 and thread 2 finish their own routine, the `signaller` thread calls `broadcast()` and wakes up all threads except itself. After that, the `signaller` tries to call `signal()`, but since there is no thread blocked by `cond_var`, `signal()` returns `false`.
```
Comment (DavidSpickett) [inline demos/conditionvariables/README.md]: "Once thread 1 and thread 2 have finished"

I think "finished" makes it clearer that they have ended, not just got to a yield point.

File: demos/conditionvariables/README.md [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+Then `signaller` creates another thread called `final_signal` and waits for `cond_var`, which shows that a thread can wait for a condition variable which has once been signalled.
```
Comment (DavidSpickett) [inline demos/conditionvariables/README.md:48]: I would add:
"This wait causes a yield to the other threads that were signalled in the `broadcast`, allowing them to finish."

File: demos/conditionvariables/README.md [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+When the `final_signal` thread calls `signal()` on `cond_var`, it finishes its execution. Since the `signaller` has no job to do, it exits after it is chosen to run and ends the demo.
```
Comment (DavidSpickett) [inline demos/conditionvariables/README.md]: "Then the `final_signal` thread calls `signal()` on `cond_var` and finishes. Next `signaller` runs, it has no work left to do so exits and ends the demo."

**Review by DavidSpickett — CHANGES_REQUESTED** [review]

File: demos/conditionvariables/README.md (hunk 10 above)
Comment (OscarShiang) [reply demos/conditionvariables/README.md:48]: Isn't that the threads exit because of the `yield()` after we call `broadcast()`?

Comment (DavidSpickett) [reply demos/conditionvariables/README.md:48]: Doh, yes you're right!

File: demos/conditionvariables/README.md [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+This function unblocks all threads blocked on the specified condition variable `cv`.
+
+`broadcast()` has a similar functionality to `signal()`. `broadcast()` will wake up all the threads blocked by the condition variable while `signal()` only wakes up one of them.
+
+## Walkthrough
+
+From starter, the `setup()` function creates 5 threads that are set to run `waiter()` and initializes a condition variable called `cond_var`. After the creation, the `setup()` thread jumps to `signaller()` function.
+
+In `signaller()`, it first yields to check if the condition variable successfully blocks all other threads. Each `waiter` thread will print out `"Waiting..."` and then wait on `cond_var` thus we'll end up backing to the `signaller` thread. Then the `signaller` thread uses `signal()` to first wake and then yields to thread 1 and thread 2. The scheduler will choose the threads woken up by `cond_var` after the `signaller` thread yields.
```
Comment (DavidSpickett) [inline demos/conditionvariables/README.md]: "end up back in the `signaller` thread"

"wake and then yield to" (no "s" on the yield)

**Review by DavidSpickett — CHANGES_REQUESTED** [review]

File: demos/conditionvariables/README.md (hunk 6 above)
Comment (DavidSpickett) [reply demos/conditionvariables/README.md]: Ping on this in case it gets hidden by Github's UI.

**Review by DavidSpickett — CHANGES_REQUESTED** [review]

**Review by DavidSpickett — APPROVED** [review]
LGTM
